src/lib/storage/dictionary_segment.hpp

```diff
@@ -25,51 +31,102 @@ class DictionarySegment : public BaseSegment {
   /**
    * Creates a Dictionary segment from a given value segment.
    */
-  explicit DictionarySegment(const std::shared_ptr<BaseSegment>& base_segment);
+  explicit DictionarySegment(const std::shared_ptr<BaseSegment>& base_segment) {
+    std::shared_ptr<ValueSegment<T>> value_segment = std::dynamic_pointer_cast<ValueSegment<T>>(base_segment);
+
+    // create dictionary
+    _dictionary = std::make_shared<std::vector<T>>(value_segment->values());
+    std::sort(_dictionary->begin(), _dictionary->end());
+    // eliminates all but the first element from every consecutive group of equivalent elements
+    auto last = std::unique(_dictionary->begin(), _dictionary->end());
+    // the dictionary must be resized following the unique operation
+    _dictionary->erase(last, _dictionary->end());
+
+    // create Attribute Vector with the most fitting width
+    if (_dictionary->size() <= std::numeric_limits<uint8_t>::max()) {
+      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint8_t>>(value_segment->size());
+    } else if (_dictionary->size() <= std::numeric_limits<uint16_t>::max()) {
+      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint16_t>>(value_segment->size());
+    } else {
+      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint32_t>>(value_segment->size());
+    }
```

Comment:
I'd propose to move the code into a function (either in the DictionarySegment or as an additional function in the FixedSizeAttributeVector.hpp). This makes it easier to find in case we later want to add smaller or larger indexes.

Reply:
That is a good idea, I really liked that in your code and it helps with the code-structure